### DISCUSSION TITLE
How to set versioning for some Docker images in order to find updates?

### DISCUSSION THREAD

**Discussion by flowertwig** [renovatebot/renovate#23038] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

In my [test repo](https://github.com/flowertwig/renovate-test), I would like renovate to suggest the following updates:
```
amazoncorretto:20.0.0-alpine3.17 -> amazoncorretto:20.0.1-alpine3.17
openjdk:21-ea-25 -> openjdk:21-ea-28
```
Onboarding PR: https://github.com/flowertwig/renovate-test/pull/1

Updates are suggested as expected for `bellsoft/alpaquita-linux-base` and `eclipse-temurin`.

### What you tried so far.

renovate.json: https://github.com/flowertwig/renovate-test/blob/c702dd7ae9f8b074c863c647d5917d06e2fcac64/renovate.json
I have tried various degrees of "specificity", e.g.
`regex:^21-ea-(?<major>\d+)$`
vs
 `regex:^(?<major>\\d+)(-(?<prerelease>ea)?)?-?(?<minor>\\d+)?-?(?<compatibility>.+)?$`
but none has worked.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: packageFiles with updates
{
  "baseBranch": "main"
  "config": {
    "dockerfile": [
      {
        "deps": [
          {
            "depName": "bellsoft/alpaquita-linux-base",
            "currentValue": "stream-musl-230524",
            "replaceString": "bellsoft/alpaquita-linux-base:stream-musl-230524",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "final",
            "updates": [
              {
                "bucket": "major",
                "newVersion": "stream-musl-230526",
                "newValue": "stream-musl-230526",
                "newMajor": 230526,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/bellsoft-alpaquita-linux-base-230526.x"
              }
            ],
            "packageName": "bellsoft/alpaquita-linux-base",
            "versioning": "regex:(?<compatibility>.+)?-(?<major>\\d+)?",
            "warnings": [],
            "registryUrl": "https://index.docker.io",
            "currentVersion": "stream-musl-230524",
            "isSingleVersion": true,
            "fixedVersion": "stream-musl-230524"
          }
        ],
        "packageFile": "alpaquita-linux-base/Dockerfile"
      },
      {
        "deps": [
          {
            "depName": "amazoncorretto",
            "currentValue": "20.0.0-alpine3.17",
            "replaceString": "amazoncorretto:20.0.0-alpine3.17",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "final",
            "updates": [],
            "packageName": "amazoncorretto",
            "versioning": "regex:^(?<major>\\d+)?\\.?(?<minor>\\d+)?\\.?(?<patch>\\d+)?-?(?<compatibility>.+)?$",
            "warnings": [],
            "registryUrl": "https://index.docker.io",
            "currentVersion": "20.0.0-alpine3.17",
            "fixedVersion": "20.0.0-alpine3.17"
          }
        ],
        "packageFile": "corretto/Dockerfile"
      },
      {
        "deps": [
          {
            "depName": "openjdk",
            "currentValue": "21-ea-25",
            "replaceString": "openjdk:21-ea-25",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "final",
            "updates": [],
            "packageName": "openjdk",
            "versioning": "regex:^21-ea-(?<major>\\d+)$",
            "warnings": [],
            "registryUrl": "https://index.docker.io",
            "currentVersion": "21-ea-25",
            "fixedVersion": "21-ea-25"
          }
        ],
        "packageFile": "openjdk-21/Dockerfile"
      },
      {
        "deps": [
          {
            "depName": "eclipse-temurin",
            "currentValue": "17.0.6_10-jre-focal",
            "replaceString": "eclipse-temurin:17.0.6_10-jre-focal",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "final",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "17.0.7_7-jre-focal",
                "newValue": "17.0.7_7-jre-focal",
                "newMajor": 17,
                "newMinor": 0,
                "updateType": "patch",
                "branchName": "renovate/eclipse-temurin-17.x"
              }
            ],
            "packageName": "eclipse-temurin",
            "versioning": "regex:^(?<major>\\d+)?\\.?(?<minor>\\d+)?\\.?(?<patch>\\d+)?_?(?<build>\\d+)?-?(?<compatibility>.+)?$",
            "warnings": [],
            "registryUrl": "https://index.docker.io",
            "currentVersion": "17.0.6_10-jre-focal",
            "isSingleVersion": true,
            "fixedVersion": "17.0.6_10-jre-focal"
          }
        ],
        "packageFile": "temurin-17/Dockerfile"
      }
    ]
  }
}
```

</details>

**Comment by rarkins** [comment]
The default `docker` versioning should handle your `amazoncorretto` case fine.

For `openjdk` you'll need to use regex versioning where you have the `21` part as major and `25` as minor

**Reply by flowertwig** [reply]
Tried removing the config for `amazoncorretto`. Log confirms I'm getting the default versioning `regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$`, but no update found.

Tried `regex:^(?<major>\\d+)-ea-(?<minor>\\d+)$` for `openjdk`, but no update found. As far as I understand the documentation, for the sake of finding an update, it shouldn't matter whether I correctly match major/minor/patch etc as long as a version is found where any of them has a greater value, though `updateType` might end up being wrong of course.

I've examined both of the above on https://regex101.com/ and they match the correct groups. Seems like I must be missing something, or this is a bug.

**Reply by rarkins** [reply]
Regex is never the default versioning, you just be configuring it somehow

**Reply by flowertwig** [reply]
Your comment made me remove  "extends": ["config:base"] which was included in the onboarding PR, and now everything is working as expected. 🤯 🥳 Many thanks for pointing me in this direction.

Now, I just need to find what in the base config is wrecking some local versioning settings, but not others. In any case, is this expected behaviour?

**Reply by rarkins** [reply]
Config base shouldn't be interfering in versioning in any way

**Reply by flowertwig** [reply]
That doesn't seem to be the case. I tried reverting to the original renovate.json that was in the onboarding PR:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base"]
}
```
Now it finds one update:
```
Update eclipse-temurin Docker tag to v17.0.7_7-jre-focal
Schedule: ["at any time"]
Branch name: renovate/eclipse-temurin-17.x
Merge into: main
Upgrade eclipse-temurin to 17.0.7_7-jre-focal
```
Then I removed config:base to have just:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": []
}
```
Now, it still finds only one update, but a different one:
```
Update amazoncorretto Docker tag to v20.0.1
Schedule: ["at any time"]
Branch name: renovate/amazoncorretto-20.x
Merge into: main
Upgrade amazoncorretto to 20.0.1-alpine3.17
```

**Reply by flowertwig** [reply]
With config base, versioning is `regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$`.
Without config base, versioning is `docker`.

**Reply by HonkingGoose** [reply]
I think the workaround for Java LTS versions is included in the `config:base` preset. So that may be causing the different behavior depending on whether you extend from `config:base`?

https://docs.renovatebot.com/presets-workarounds/#workaroundsjavaltsversions

**Reply by flowertwig** [reply]
Looks like it. I tried the "resolved" version of `config:base` from https://docs.renovatebot.com/presets-config/#configbase, but without `workarounds:all`, and getting docker versioning on all images as expected.

**Reply by flowertwig** [reply]
Confirmed. This is a workaround:
```
"ignorePresets": [
    "workarounds:javaLTSVersions"
  ]
```

**Reply by rarkins** [reply]
Thanks for debugging it.

Do I understand right that you have `openjdk:21-ea-25` as a tag, which is not compatible with preset `workarounds:javaLTSVersions`?

**Reply by flowertwig** [reply]
Yes. It matches `ea-25` as `compatibility`.

**Reply by flowertwig** [reply]
In my opinion, limiting java updates to only LTS should be opt-in. It's not like they are "unstable" or "pre-release".

**Reply by viceice** [reply]
> In my opinion, limiting java updates to only LTS should be opt-in. It's not like they are "unstable" or "pre-release".

no, renovate does that for all other tools too, like nodejs. most users prefer to stick on lts version.

**Reply by flowertwig** [reply]
As far as I understand, the debug log is incorrect since the versioning logged is always the one that I have configured in my `renovate.json`, regardless of the one that is actually applied.
**With `workarounds:javaLTSVersions`**
```
"deps": [
          {
            "depName": "openjdk",
            "currentValue": "21-ea-25",
            "replaceString": "openjdk:21-ea-25",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "final",
            "updates": [],
            "packageName": "openjdk",
            "versioning": "regex:^(?<major>\\d+)(-(?<prerelease>ea))?-(?<patch>\\d+)?-?(?<compatibility>.+)?$",
            "warnings": [],
            "registryUrl": "https://index.docker.io",
            "currentVersion": "21-ea-25",
            "fixedVersion": "21-ea-25"
          }
        ],
        "packageFile": "openjdk-21/Dockerfile"
      }
```
**Without `workarounds:javaLTSVersions`**
```
"deps": [
          {
            "depName": "openjdk",
            "currentValue": "21-ea-25",
            "replaceString": "openjdk:21-ea-25",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "final",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "21-ea-28",
                "newValue": "21-ea-28",
                "newMajor": 21,
                "newMinor": 0,
                "updateType": "patch",
                "branchName": "renovate/openjdk-21.x"
              },
              {
                "bucket": "major",
                "newVersion": "22-ea-3",
                "newValue": "22-ea-3",
                "newMajor": 22,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/openjdk-22.x"
              }
            ],
            "packageName": "openjdk",
            "versioning": "regex:^(?<major>\\d+)(-(?<prerelease>ea))?-(?<patch>\\d+)?-?(?<compatibility>.+)?$",
            "warnings": [],
            "registryUrl": "https://index.docker.io",
            "currentVersion": "21-ea-25",
            "isSingleVersion": true,
            "fixedVersion": "21-ea-25"
          }
        ]
```

**Reply by viceice** [reply]
it's because the workaround is applied via package rule, so it overrides any global option

**Reply by flowertwig** [reply]
Yes, but shouldn't the overriding versioning be reflected in the debug log? Logging one versioning and applying a different one is extremely confusing.

**Reply by viceice** [reply]
that's complicated because not all config options are copied to the update object on debug logs. they get applied after the log message.

**Reply by rarkins** [reply]
So "packageFiles with updates" shows the wrong versioning?

**Reply by flowertwig** [reply]
Yes. It shows the versioning from my local config, regardless of whether `workarounds:javaLTSVersions` is ignored.

**Reply by rarkins** [reply]
Is this reproduced in a public repo still which I can test against?

**Reply by flowertwig** [reply]
Yup: https://github.com/flowertwig/renovate-test
Local versioning config for openjdk, but not for temurin.
`workarounds:javaLTSVersions` is not ignored, so I would expect debug log to show same versioning for openjdk and temurin in `packageFiles with updates`. However, only temurin shows versioning from the preset, while openjdk shows the one from the local config.

**Reply by rarkins** [reply]
I see:

```
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "extends": ["config:base"],
         "packageRules": [
           {
             "matchPaths": ["**/openjdk-21/**"],
             "versioning": "regex:^(?<major>\\d+)(-(?<prerelease>ea))?-(?<patch>\\d+)?-?(?<compatibility>.+)?$",
             "ignoreUnstable": false
           }
         ]
       }
```

then:

```json
{
         "dockerfile": [
           {
             "deps": [
               {
                 "depName": "openjdk",
                 "currentValue": "21-ea-25",
                 "replaceString": "openjdk:21-ea-25",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "final",
                 "updates": [],
                 "packageName": "openjdk",
                 "versioning": "regex:^(?<major>\\d+)(-(?<prerelease>ea))?-(?<patch>\\d+)?-?(?<compatibility>.+)?$",
                 "warnings": [],
                 "registryUrl": "https://index.docker.io",
                 "currentVersion": "21-ea-25",
                 "fixedVersion": "21-ea-25"
               }
             ],
             "packageFile": "openjdk-21/Dockerfile"
           },
           {
             "deps": [
               {
                 "depName": "eclipse-temurin",
                 "currentValue": "17.0.6_10-jre-focal",
                 "replaceString": "eclipse-temurin:17.0.6_10-jre-focal",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "final",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "17.0.7_7-jre-focal",
                     "newValue": "17.0.7_7-jre-focal",
                     "newMajor": 17,
                     "newMinor": 0,
                     "updateType": "patch",
                     "branchName": "renovate/eclipse-temurin-17.x"
                   }
                 ],
                 "packageName": "eclipse-temurin",
                 "versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
                 "warnings": [],
                 "registryUrl": "https://index.docker.io",
                 "currentVersion": "17.0.6_10-jre-focal",
                 "isSingleVersion": true,
                 "fixedVersion": "17.0.6_10-jre-focal"
               }
             ],
             "packageFile": "temurin-17/Dockerfile"
           }
         ]
       }           
```

The package rule matches only the first

**Reply by flowertwig** [reply]
That is intentional since temurin is only there as a comparison.
However, I now also added `allowedVersions` that accepts 21 as major to the openjdk config, and it appears that it was actually `"allowedVersions": "/^(?:8|11|17)(?:\\.|-|$)/"` from `workarounds:javaLTSVersions` that caused it not to find any updates.
**I.e. my assumption that the `versioning` was logged incorrectly is probably wrong.**
The only "bug" here is likely that the `workarounds:javaLTSVersions` `versioning` regex is not compatible with openjdk.
To avoid confusion, it might also be a good idea to consider debug logging which presets are in effect and what configuration that they apply, in this case `allowedVersions`.
